### PR TITLE
[AIRFLOW-1166] Speed up _change_state_for_tis_without_dagrun

### PR DESCRIPTION
**Pull request by bolkedebruin**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1166

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:

_change_state_for_tis_without_dagrun was locking a significant
amount of tasks uncessarily. This could end up in a deadlock
in the database due to the time the lock stood.

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Unit tests added. 

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

@aoen @saguziel

### CODE REVIEW

**Comment by bolkedebruin** [comment]
@aoen @artwr @saguziel I'm pretty sure you guys must see this issue, except when the previous code is not in your fork yet.

errors on travis are unrelated.

**Comment by bolkedebruin** [comment]
@saguziel @aoen please review. I would really like to get it in 1.8.2 and besides I'm pretty sure you will (eventually) see this issue.

File: tests/jobs.py [hunk 1]
```diff
@@ -613,6 +613,77 @@ def test_concurrency(self):
...
+        ti.refresh_from_db(session=session)
+        self.assertEqual(ti.state, State.SCHEDULED)
+
+        ti2.refresh_from_db(session=session)
+        self.assertEqual(ti2.state, State.SCHEDULED)
+
+        dr.refresh_from_db(session=session)
+        dr.state = State.FAILED
+
+        # why o why
```
Comment (aoen) [inline tests/jobs.py:672]: ?

**Comment by aoen** [comment]
LGTM

File: tests/jobs.py (hunk 1 above)
Comment (bolkedebruin) [reply tests/jobs.py:672]: It is just strange why I need to merge in the current session while DagRun should be part of that session.

**Pull request closed without merging**